compiler/src/dotty/tools/dotc/typer/Implicits.scala

```diff
@@ -928,9 +933,70 @@ trait Implicits:
           if normalImports.isEmpty then suggestedImports else normalImports
         end hiddenImplicitsAddendum
 
-        msg(userDefined.getOrElse(
-          em"no implicit argument of type $pt was found${location("for")}"))() ++
-        hiddenImplicitsAddendum
+
+        /** Extracting the message from a method parameter, e.g. in
+         *
+         *  trait Foo
+         *
+         *  def foo(implicit @annotation.implicitNotFound foo: Foo)
+         */
+        def userDefinedParamMessage = paramSymWithMethodTree.flatMap { (sym, tree) =>
+          userDefinedMsg(sym, defn.ImplicitNotFoundAnnot).map { rawMsg =>
+            val paramSyms = typeVariablesInScope(sym)
```

Comment:
"type variables" in the compiler usually refer the variables the compiler internally creates when performing type inference so it's better to not use that term, `typeParamsInScope` would be fine, and since this method is only used once you could inline it here actually: `val typeParamsInScope = ...`